### ISSUE TITLE
BranchUniversalObject.generateShortUrl() does not always run callback

### ISSUE THREAD

**Issue by eirikwah** [BranchMetrics/android-branch-deep-linking#517]
When initializing Branch with `Branch.getAutoInstance(applicationContext);` and then calling `BranchUniversalObject.generateShortUrl(Context, LinkProperties, BranchLinkCreateListener);` before the initialization is finished, the callback `BranchLinkCreateListener` will never be called.

We get this bug in our application when users trigger a link generation just after the app has been installed and run for the first time. If the link generation is done for example after 3 seconds, the callback is triggered as expected.

To reproduce, put this code in a button click handler (or another place it can be tested with ease):

```java
// Example to reproduce the bug:

Branch.getAutoInstance(mApplicationContext); // This code is not equal to our production code, as we call this method another place. But we put this call here to be able to reproduce this issue.

BranchUniversalObject branchUniversalObject = new BranchUniversalObject()
        .setCanonicalIdentifier("invitation_from/user123")
        .setTitle("Subject")
        .setContentDescription("description")
        .setContentIndexingMode(BranchUniversalObject.CONTENT_INDEX_MODE.PRIVATE)
        .addContentMetadata("custom_key", "our value");

LinkProperties linkProperties = new LinkProperties()
        .setFeature("invite")
        .setChannel("someChannel");

branchUniversalObject.generateShortUrl(
        mApplicationContext,
        linkProperties,
        new Branch.BranchLinkCreateListener() {
            @Override
            public void onLinkCreate(String url, BranchError error) {
                Log.d("BranchTest", "BranchLinkCreateListener.onLinkCreate was triggered");
            }
        });
```

In the above code, `BranchLinkCreateListener.onLinkCreate was triggered` is never logged.

If we introduce a delay between the call to `getAutoInstance` and `generateShortUrl`, then the call to `generateShortUrl` works as expected:

```java
// Working example:

Branch.getAutoInstance(mApplicationContext); // This code is not equal to our production code, as we call this method another place. But we put this call here to be able to reproduce this issue.

BranchUniversalObject branchUniversalObject = new BranchUniversalObject()
        .setCanonicalIdentifier("invitation_from/user123")
        .setTitle("Subject")
        .setContentDescription("description")
        .setContentIndexingMode(BranchUniversalObject.CONTENT_INDEX_MODE.PRIVATE)
        .addContentMetadata("custom_key", "our value");

LinkProperties linkProperties = new LinkProperties()
        .setFeature("invite")
        .setChannel("someChannel");

Handler handler = new Handler();
// Call with a 3 second delay. (A zero second delay will NOT work):
handler.postDelayed(() -> {
    branchUniversalObject.generateShortUrl(
            mApplicationContext,
            linkProperties,
            new Branch.BranchLinkCreateListener() {
                @Override
                public void onLinkCreate(String url, BranchError error) {
                    Log.d("BranchTest", "BranchLinkCreateListener.onLinkCreate was triggered");
                }
            });
}, 3000);
```

In the above code, `BranchLinkCreateListener.onLinkCreate was triggered` *IS*  logged.

Tested on version 2.14.1.

**Comment by sojanpr** [comment]
@eirikwah  `generateShortUrl()` will return a link if Branch is already initialized. If not it wait until Branch init finishes and generate the url after that. We are keeping only a weak reference to all callbacks in our SDK in oder to avoid any memory leaks. Is there  a chance the callback object get killed up on Branch init session finished.

Adding the following code in the `MainActivity` adde to the testbed in this repo is working fine.
It 

```
@Override
protected void onStart() {
    super.onStart();
    branch = Branch.getInstance();
    branchUniversalObject.generateShortUrl(this, new LinkProperties(), new Branch.BranchLinkCreateListener() {
        @Override
        public void onLinkCreate(String url, BranchError error) {
            Log.i("Callback_Test", “Link callback returned " +url + " error is" + error);
        }
    });
    branch.initSession(new Branch.BranchUniversalReferralInitListener() {
        @Override
        public void onInitFinished(BranchUniversalObject branchUniversalObject, LinkProperties linkProperties, BranchError error) {
           Log.i("Callback_Test", “Init finished");
        }
    }, this.getIntent().getData(), this);
}
```
Output is
```Callback_Test: Init finished
Callback_Test: Link callback returned https://bnctestbed.test-app.link/BW84h8oxDI
```

**Comment by eirikwah** [comment]
Thank you for following up this, @sojanpr.

If I read the documentation correctly, `Branch.getAutoInstance(Context)` should trigger initialization (of the session) automatically, and this is the method we use (see my example in my first post).

Is it a requirement that the initialization must happen *after* `generateShortUrl()`?

Always working (in testbed app):

```java
@Override
protected void onStart() {
    super.onStart();
    branchUniversalObject.generateShortUrl(this, new LinkProperties(), new Branch.BranchLinkCreateListener() {
        @Override
        public void onLinkCreate(String url, BranchError error) {
            Log.i("Callback_Test", “Link callback returned " +url + " error is" + error);
        }
    });
   Branch.getAutoInstance();
}
```

Sometimes not working (in testbed app):

```java
@Override
protected void onStart() {
    super.onStart();
    Branch.getAutoInstance();
    branchUniversalObject.generateShortUrl(this, new LinkProperties(), new Branch.BranchLinkCreateListener() {
        @Override
        public void onLinkCreate(String url, BranchError error) {
            Log.i("Callback_Test", “Link callback returned " +url + " error is" + error);
        }
    });
}
```

Maybe related: If I switch the sequence in your example, so that `branch.initSession` is called before `generateShortUrl`, then the log will show "Link callback returned", but not always "Init finished". Is this expected?

I write "sometimes" and "not always", because I cannot reproduce this (assumed) bug all the time. Can this be timing-related?

PS: Calling `Branch.getAutoInstance(this);` in `Application.onCreate` instead of in `MainActivity.onStart()` seems to always work for the testbed app. But this does not fix the situation for our app running on "LGE Nexus 5 (Android 6.0.1, API 23)". This is a very hard bug to reproduce (when `getAutoInstance` is called from `Application.onCreate`), as we can only recreate this on the very first launch after downloading the app from the Google Play Store. I am guessing that the async parts of `getAutoInstance` takes longer time to run the first time the app is started after being installed, and that is why I reproduce this (assumed) bug by moving the call `getAutoInstance` from `Application.onCreate` to `MainActivity.onStart()`.

**Comment by sojanpr** [comment]
@eirikwah As you have seen in the documentation we suggest to call `Branch.getAutoInstance(this);` in `Application#onCreate()`.  You can call `generateShortUrl()` anytime but it will be returned only after Branch SDK is initialized. SDK init happens on the lifecycle event `onResume()` for the very first activity.

Our awesome integration team can help you here more. Can you please write to integrations@branch.io for a quicker response

**Comment by eirikwah** [comment]
@sojanpr We *are* calling `Branch.getAutoInstance(this);` in `Application#onCreate()` in our production code. This works in 99.9% of the cases, but sometimes the listener for `generateShortUrl()` does not get called.

Thanks for the suggestion to contact integrations@branch.io. However I do not need support, I am only trying to report what I believe may be a bug.

Sorry for not being able to give you a 100% reliable way of reproducing our problem. I hope you can take the time to try to reproduce this on your side - even if you may have reason to believe that this is not a bug in Branch, but an error on our side.

In any case - we continue to use Branch. And for the 0.1 % of our users that don't get a link we have now created a workaround (we detect that there is no response from the Branch API within the timeout, and use another link instead).

**Comment by sojanpr** [comment]
@eirikwah Really appreciate your help and the above report. We are trying to reproduce this case and  will fix it asap if this an issue with Branch. We are sorry that you have this issue but do you think this has any dependency on network latency. We have seen some issue where the callback is delayed for a long time when there is some network issues
Thank you very much for using Branch and will update you soon on this issue

**Comment by eirikwah** [comment]
Thanks again. I have tested without network connection, and with (simulated) 100 % package loss on the network, but not been able to reproduce this in those cases. When having 100 % package loss, the network connection times out after the timeout value and retry count we configure Branch to use (10 seconds, 3 retries). So that part works as it should.

The physical device that always fails on the first launch after installation ("LGE Nexus 5 (Android 6.0.1, API 23)"), is connected to a good wifi. So I am guessing this is probably not related to network latency.

**Comment by sojanpr** [comment]
@eirikwah Thanks for the details. Will keep you updated

**Comment by sojanpr** [comment]
@eirikwah  Could you please help us on reproducing it. Can you please run the TestBed app on the same device with the same config as of your app and see if you can reproduce the issue?

**Comment by sojanpr** [comment]
@eirikwah We have found one corner case  where the  callback may not be called. We have fixed this issue with v2.14.5 can you please give it a try
